Convert `SimulationState` and `SimulationInputs` to dataclasses

LiamPattinson opened the pull request:
Additionally:

- Adds a `MagneticProfile` dataclass, which replaces the dictionary previously in use.
- Renames files and moves items between them.
- Adds dataclass for outputs

This is mostly work undertaken last year by @ZedThree; all I've done is check through it and tinker around the edges. The primary aim of this PR is to make progress towards improving the user interface.

Fixes one of the points in https://github.com/FusionDLS/FusionDLS/issues/34 and some points in https://github.com/FusionDLS/FusionDLS/issues/3 (although that issue might now be out-of-date).

---

Some further work is needed on the outputs, but I've raised this in its current state to avoid this PR becoming too bloated. I'd prefer to raise separate issues/PRs to make those improvements at a later date. However, if you have any concerns/comments about the changes made so far, absolutely feel free to let me know!

ZedThree committed:
Use more accurate physical constants from scipy

ZedThree committed:
Convert `SimulationInputs` to a dataclass

ZedThree committed:
Convert `SimulationState` to `dataclass`

ZedThree committed:
Move `radios` into main `SimulationInputs` class

ZedThree committed:
Add `MagneticGeometry` dataclass

ZedThree committed:
Fix some `SimulationInputs` docstrings

ZedThree committed:
Move `MagneticGeometry` to new file

ZedThree committed:
Make some `MagneticGeometry` attributes properties

ZedThree committed:
Move `refineGrid` into `MagneticGeometry`

ZedThree committed:
Rename `LRBv21` -> `solver`

ZedThree committed:
Move `Iterate` into `solver`

ZedThree committed:
Add dataclass for solver outputs

ZedThree committed:
Check for allowable `control_variable` inputs

LiamPattinson committed:
Lint and format

LiamPattinson committed:
Add support for Python 3.13 in test workflow

LiamPattinson committed:
Update linting to Python 3.10

LiamPattinson committed:
Reformat SimulationInput/State docs

LiamPattinson committed:
Make _drop_properties classmethod of MagneticGeometry

Future-proofing in case other properties are added

LiamPattinson committed:
Edit grid refinement docs

LiamPattinson committed:
Use dynamic grids by default

LiamPattinson committed:
Make SimulationOuptut a Mapping type

Allows it to be manipulated as an immutable dict.

LiamPattinson committed:
Update notebook 1

LiamPattinson committed:
Add __setitem__ to SimulationOutputs

LiamPattinson commented:
Fixes https://github.com/FusionDLS/FusionDLS/issues/34

Some further changes that didn't seem worth opening a new PR for:

- Use dynamic grids by default, now need to set `static_grid = True` to disable.
- `SimulationOutputs` is now a `Mapping` type, giving it more `dict`-like features (the most important being `.keys()` and `.items()`). I could have made it a `MutableMapping` type, but that would involve adding `__delitem__`, and allowing users to call `.pop(key)` on a `dataclass` just doesn't feel right. I added `__setitem__`, so you can still modify it via `output[key] = val`. (see [this table](https://docs.python.org/3/library/collections.abc.html#collections-abstract-base-classes) for more info). 
- Further updates to docstrings etc
- Updated the first notebook. Should I update the others too, or should that wait until https://github.com/FusionDLS/FusionDLS/issues/42?

mikekryjak commented:
Thanks Liam, can you wait with the examples? I will try to update them today. I haven't reviewed the rest of this PR yet.

mikekryjak commented:
I have a request on this one - can we take the inputs overhaul a bit further? In my mind, there is no need for a separation between inputs directly passed to `run_dls` and to the `constants` dictionary. I think it would be much clearer if they were all in `SimulationInputs`, so that the user has only one place to look for settings. This would also make it easy to add the capability for accepting input files that we planned to implement. 

In fact, I made the `SimulationInputs` object as an intermediate step towards this exactly kind of system - it merged the DLS function inputs, radios and constants into one object. The reason I didn't go all the way is because the refactor was already extensive and I thought it would have been too risky. 

Another thing is that the inputs contain variables that define the domain, but they really should be packaged in their own object instead. I created the `Profile` object just for that purpose, but again I didn't fully integrate it with the main code because I wanted to constrain the scope of my refactor. It contains all of the geometry/magnetic profile information:
https://github.com/FusionDLS/FusionDLS/blob/8b618a96eae63001c4d0e60f83a30682fbb79c9e/src/fusiondls/Profile.py#L17-L31

It also contains lots of methods for manipulating the profiles and calculating their various properties.

@cydcowley @ZedThree @LiamPattinson what do you think of this idea? Do you think this should be done in this PR or a separate one?

LiamPattinson commented:
I think that sounds sensible -- I was hoping to sneak in a rework of the `constants` dict before the merge anyway.

Should the args in the constants dict be turned into plain arguments to `run_dls`? Or is there some other intermediate settings object that should replace it?

Looking at `Profile.py`, it looks like there's a lot of overlap with the `MagneticGeometry` object that @ZedThree introduced. Should we aim to port features from `Profile` over to there, or should it go the other way around?

mikekryjak commented:
Not sure if I understand your first question. I would like for everything in `constants` and the `run_dls` arguments to be put into `SimulationInputs`, so that you can do `run_dls(SimulationInputs)`, and `constants` or `radios` no longer exist. What do you think? Alternatively we could unpack all of `SimulationInputs` into individual arguments and then pack them into a `SimulationInputs` in the first line of `run_dls`, I guess.

On the second point: I just reviewed `MagneticGeometry` for the first time. I have to say that I am not sure if I fully understand the intent behind it, i.e. whether @ZedThree intended for it to replace `Profile` or not.

I like how it includes the properties and key functionality of both `Profile` and `refineGrid` and it would make sense to implement the rest of `Profile` methods within it. But I would like to make sure I understand Peter's intent behind this refactor properly first. There is also the question of what to do with `Morph`, which is a class that effectively interpolates between two `Profiles`.

LiamPattinson commented:
> Not sure if I understand your first question. I would like for everything in `constants` and the `run_dls` arguments to be put into `SimulationInputs`, so that you can do `run_dls(SimulationInputs)`, and constants or radios no longer exist. What do you think? 

Okay, I get you now. I would normally lean towards having a bit more structure in the input settings, so instead of having all settings on a flat level within `SimulationInputs` you would instead have something like:

```python

@dataclass
class SimulationInputs:
    """Just an example, names and the exact breakdown could be anything really"""
    profile: MagneticGeometry  # Or 'Profile'?
    constants: Constants  # There's probably a better name for this
    grid_settings: GridSettings  # For refinement_ratio etc
    ...
    verbosity: int = 0  # Can also have plain constants

@dataclass
class MagneticGeometry:
    """To be determined from a SOLPS/GEQDSK file"""
    Bpol: FloatArray
    Btot: FloatArray
    ...

@dataclass
class Constants:
    """To replace the 'constants' dict"""
    gamma_sheath: float
    Tt: float
    ...
```

This would bring a few advantages:

- It's easier to do parameter scans using the smaller objects, especially if they contain methods designed to assist with it
- It's easier for users to figure out how to build a few small objects than to figure out how to fill in every field of `SimulationInputs` in one go.
- The structure of a future input file could match the structure of the code:

```toml
# TOML file used as input file
# global settings go here
verbosity = 0

[constants]
gamma_sheath = 7.0
Tt = 0.5
# ...

[grid-settings]
refinement-ratio = 5.0
refinement-width = 1.0
# ...
```

(There's a library called [`pyserde`](https://yukinarit.github.io/pyserde/guide/en/) that can automate the generation of input-file readers from this sort of setup, but I haven't tested it)

In the first instance however, I can merge everything into `SimulationInputs` and refine things in a later PR.

> Alternatively we could unpack all of `SimulationInputs` into individual arguments and then pack them into a `SimulationInputs` in the first line of `run_dls`, I guess.

I feel like that would result in a very complex interface to `run_dls`, and it would be easier to understand if there were fewer arguments to pass in.

ZedThree commented:
> On the second point: I just reviewed `MagneticGeometry` for the first time. I have to say that I am not sure if I fully understand the intent behind it, i.e. whether @ZedThree intended for it to replace `Profile` or not.

I think I probably intended it to eventually replaced `Profile`, but I didn't want to disturb it before I fully understand what `Profile` was doing exactly

LiamPattinson committed:
Add grid_resolution to run_dls

Renames refinement_* args to grid_refinement_*

LiamPattinson committed:
Merge branch 'main' into state-dataclasses

LiamPattinson committed:
Add cooling_curves dict, allowing selection from str

Users don't have to import AnalyticCoolingCurves, and can
instead pass in their choice of curve as a string arg.

LiamPattinson committed:
Rename AnalyticCoolingCurves.py to analytic_cooling_curves.py

https://peps.python.org/pep-0008/#package-and-module-names

mikekryjak commented:
Hi @LiamPattinson you raise a good point, I made an omission in my earlier post. I would like it to look like `run_dls(SimulationInputs, MagneticGeometry)` or something similar. The geometry must be separate for sure, it's more logical this way and as you say the geometry has associated methods.

Meanwhile I think there are really not that many other options, whether they are process conditions or solver settings. I think it would be simplest for them all to live in SimulationInputs for now, with no sub-containers like `constants` inside.

LiamPattinson committed:
Add interpolated B() function to MagneticGeometry

LiamPattinson committed:
Remove constants dict arg to run_dls

All folded into SimulationInputs, which the user must build
and pass in themselves. Magnetic profile args have been
moved from SimulationInputs, and a MagneticGeometry object
must now be passed around too.

LiamPattinson committed:
Format tests

LiamPattinson commented:
I've made a few more changes which I hope is closer to what you're looking for. The example notebooks haven't been updated, but `test_analytic.py` shows the new interface.

- Remove geometry args from `SimulationInputs` -- no need to duplicate them.
- Pass `SimulationInputs` and `MagneticGeometry` into `run_dls`, remove most other args.
- `SimulationInputs` is no longer edited after being passed into `run_dls` (`MagneticGeometry` isn't either).

A further change is that the cooling curve can be specified with a string, so the user doesn't have to import `LFuncX`.

mikekryjak commented:
Thanks @LiamPattinson this looks good. I also like how you left `verbosity` and `diagnostic_plot`, I think that makes sense.
 A couple of comments:

- I think having a function input for the cooling curve is better, because it A) teaches the user that it's easy to supply your own curve, and B) it forces them to give a bit of thought to what curve they're using. The results are very sensitive to cooling curve choice but this is something that's too often treated like a black box by physicists.
- I noticed that `solver.py` has no git history. Was it renamed without doing `git mv`?

ZedThree commented:
> * I think having a function input for the cooling curve is better, because it A) teaches the user that it's easy to supply your own curve, and B) it forces them to give a bit of thought to what curve they're using. The results are very sensitive to cooling curve choice but this is something that's too often treated like a black box by physicists.

Having it as a string makes it easier to supply via an input file or the command line, but it should also be possible to supply an actual function instead, just needs some checking to see if it's `str` or not.

> * I noticed that `solver.py` has no git history. Was it renamed without doing `git mv`?

Might just be the PR being weird? I renamed it in [19d5545](https://github.com/FusionDLS/FusionDLS/pull/39/commits/19d55456b6c60c19c03fec668972d992405a443a)

LiamPattinson committed:
Set cooling_curve via function or string

LiamPattinson commented:
I've added the ability to set `cooling_curve` via `str` or `Callable[[float], float]` using the probably-overkill method of [descriptor fields](https://docs.python.org/3/library/dataclasses.html#descriptor-typed-fields), and updated the docs to make it clear that this is an important thing for the users to get right.

LiamPattinson committed:
Port changes from PR #45 to PR #39

Improves target heat flux control

LiamPattinson committed:
Merge branch 'main' into state-dataclasses

LiamPattinson commented:
I've now brought over the changes from https://github.com/FusionDLS/FusionDLS/pull/39. Is there anything else that needs to be done this PR?

mikekryjak commented on src/fusiondls/solver.py:
```diff
@@ -0,0 +1,857 @@
+
+    Includes points between front location and X-point."""
+
+    SparFront: FloatArray = field(init=False)
```
SparFront is the current detachment front location. I think it should be a float.

mikekryjak replied:
@LiamPattinson this is the only thing I found that I thought wasn't right.

mikekryjak committed:
Add missing docstrings

mikekryjak commented:
@ZedThree @LiamPattinson some comments:

- Is there any benefit to having `SimulationState` and `SimulationInputs` in `solver.py` instead of a separate file?
- At the moment, the PR half-implements MagneticGeometry without all of the methods of Profile, in particular the simple methods to obtain statistics as well as the more complex methods for arbitrarily adjusting the profile shape. Honestly, ideally this would be a separate PR, and probably a test of the functions of Profile would be useful for that. If we merge it like it is now then technically the Profile workflow is untouched, so maybe we should just do that and open a separate PR.
- What is `typing.py`?

LiamPattinson commented:
> Is there any benefit to having `SimulationState` and `SimulationInputs` in `solver.py` instead of a separate file?

It makes it easier to quickly check which variables are in which class when you're looking through the functions in `solver.py`, but that's about it. I think it makes sense to keep `SimulationState` there as it's quite closely tied to `run_dls`, but `SimulationInputs` could be moved to another file.

> At the moment, the PR half-implements `MagneticGeometry` without all of the methods of `Profile`, in particular the simple methods to obtain statistics as well as the more complex methods for arbitrarily adjusting the profile shape. Honestly, ideally this would be a separate PR, and probably a test of the functions of Profile would be useful for that. If we merge it like it is now then technically the Profile workflow is untouched, so maybe we should just do that and open a separate PR.

I agree that it should have been its own PR, and I'd like to merge the two into one class soon (probably as a precursor to working on the equilibrium files). The reason I've left the two separate is simply because that's how I found it when taking over this branch, and I've been trying to avoid adding too much further bloat before it gets merged.

> What is `typing.py?`

That's just a place to define some useful type aliases that can be used throughout the project. Taking the example of `PathLike`, it's quite common for functions that operate on file paths to accept either a `str` or `pathlib.Path`, and it might also work with a `pathlib.PosixPath` or `pathlib.PurePath` etc. To simplify the type hinting, it makes sense to create a small alias that can be used anywhere.

For `Scalar`, that's useful for getting around common type errors raised by static type analysis tools like MyPy or Pyright when you're working with NumPy. A lot of NumPy functions will return things like `np.float64`, which results in a type error if you pass it to a function expecting a `float`, even though they behave almost identically. If the function expects `Scalar` instead, it'll be more permissive when used with NumPy code.

mikekryjak commented:
Okay, I agree with everything you said. I've not really worked with Python typing before so this is useful to learn about. 
I'm just finishing the examples PR this morning, how about we merge that, then update the examples here and merge?

LiamPattinson committed:
Move SimulationInputs to settings.py

LiamPattinson committed:
Merge branch 'state-dataclasses' of github.com:FusionDLS/FusionDLS into state-dataclasses

LiamPattinson committed:
Reduce Ttol default to 1e-3

LiamPattinson committed:
Merge branch 'main' into state-dataclasses

LiamPattinson commented:
I've merged in the relevant changes to https://github.com/FusionDLS/FusionDLS/pull/50 and caught up to `main`. Is anything else needed for this PR?

mikekryjak commented:
@LiamPattinson just updating the examples and that thing I found in one of the comments above.

LiamPattinson commented:
> just updating the examples

Oh yeah, I'll get on that now

>  ...and that thing I found in one of the comments above.

I had a flick through and can't quite tell what you mean. Please could you point me in the right direction?

mikekryjak commented:
It was a comment in my review, maybe it's not showing for you because the review is not complete.

I'm pretty sure this should be a float. It's the parallel front location, not an array.
https://github.com/FusionDLS/FusionDLS/blob/167b80d22a86bbbe1d657f0c85a771d46172ef62/src/fusiondls/solver.py#L46

LiamPattinson committed:
Switch SparFront typehint from FloatArray to float

LiamPattinson committed:
Update basic_use notebook

LiamPattinson committed:
Linting/Formatting

LiamPattinson committed:
update analytic_test notebook

N.B. The predicted convergence at the end of the notebook is no
longer correct!

LiamPattinson commented:
I've updated `basic_use` and `analytic_test`, but ran into issues with both:

- The plots in `basic_use` are now at a lower resolution, but I get the same plots if I generate them in `main`.
- The convergence predicted in `analytic_test` is incorrect, but again, I get the same results on this branch or `main`. Has something broken somewhere?
- ~~I get some confusing NumPy errors if I don't copy `SparFront` into each new settings object, suggesting that it's being modified somewhere that it shouldn't be. Can't quite track down where it's happening.~~ [EDIT: I'm unable to replicate this error, so this was probably due to a broken Python env or something]

I'll have to get to `geometry_manipulation` tomorrow morning.

LiamPattinson committed:
Update geometry_manipulation notebook

LiamPattinson committed:
Test notebooks in CI

LiamPattinson committed:
Fix linting/formatting issues

LiamPattinson commented:
I've finished fixing up the notebooks, and I've added a test to the CI that will help us catch any divergence between the code and the example notebooks. It will fail the tests if any notebooks raise any errors, but it doesn't check that the notebooks are actually producing the correct answers. We could add a few `assert` statements so that the notebooks will also test for correctness (or better yet, `np.testing.assert_*` functions).

LiamPattinson committed:
Fix analytic_test notebook, add test to the notebook

LiamPattinson committed:
Assert specific analytic result from analytic_test notebook

mikekryjak commented:
Thanks @LiamPattinson, just to document our discussion:

- The issue with `basic_use` not showing the correct number of points is my fault, I think I generated/saved the notebook with 10 points but then decided 5 is better and didn't regenerate.
- The issue with the analytical test is because it needs `front_sheath=false` as per https://github.com/FusionDLS/FusionDLS/pull/45/commits/f3329a8eac3b49eeb31da9cab3799c20da324037

LiamPattinson commented:
Cheers, I think I've got it all fixed up now. The results in the `analytic_test` now match the previous results with only minor changes. I've also updated the last couple of cells to give an example of adding tests directly to the notebook. You can run the notebook tests using:

```bash
pip install nbmake  # pytest plugin for notebook testing
pytest --nbmake docs/examples/*.ipynb
```

LiamPattinson committed:
Revert commits until "Update basic_use notebook"

This reverts commits:

- d1d5422
- 2b6d3b3
- 6df0e72
- 930b456
- 610aac8
- 71165b5
- d1683b8
- fef723c

LiamPattinson committed:
Merge branch 'main' into state-dataclasses

Updates test workflow to include both tests and notebooks,
plus adds coverge stats

LiamPattinson committed:
Update basic_use notebook

LiamPattinson committed:
Update geometry_manipulation notebook

LiamPattinson committed:
Lint notebooks in CI

LiamPattinson committed:
[TESTS FAIL] Update analytic_test notebook

LiamPattinson committed:
Fix minor formatting issue

LiamPattinson committed:
Fix linting workflow

LiamPattinson committed:
Fix coverage in test workflow

LiamPattinson commented:
I've updated the notebooks to match the changes in https://github.com/FusionDLS/FusionDLS/pull/55. The easiest way to handle it was to revert the last few commits, merge with main, and build the changes back up from scratch.

I've also updated the testing workflow to include both the `tests` folder and the notebooks, and I've updated the linting workflow to catch the notebooks. I added test coverage reporting too, but, as is tradition, I can get it to work locally but not in the CI (I don't think I've ever gotten coverage working on my first try...).

Unfortunately, the `analytic_test` notebook is failing tests on this branch. It converges within 1%, but not the 0.5% required by the test. I don't know what has changed internally to cause this error. I'll have to look into it.

LiamPattinson committed:
Fix test coverage, attempt 2

LiamPattinson committed:
Change tolerance on analytic_test notebook

The difference in converged values compared to the main
branch is accounted for by the change from mi=3e-27 and
e=1.6e-19 to more accurate values using scipy.constants

LiamPattinson commented:
I figured out why that test was 'failing': you can replicate the previous results by setting `deuterium_mass=3e-27` and `elementary_charge=1.6e-19`. These were changed to more accurate values back in 4cc7300, so I think it's best to slightly relax the tolerance of the test and use the new results.

I think this is now good to go!

(Edit: I also got test coverage working. If you check any of the completed testing jobs, they should tell you how much of each Python file is covered by the test suite. We could now link this project to [CodeCov](https://about.codecov.io/), which reports how each pull request affects test coverage before merging, but that's very much optional)

mikekryjak reviewed (approved):
Alright, let's do this!!! 😄
